lib/cn/csched_sp3.c

```diff
@@ -1931,27 +1936,64 @@ sp3_tree_shape_check(struct sp3 *sp)
                 lsiz = pcap;
                 lclen = cn_ns_clen(&tn->tn_ns) >> 20;
             }
-        } else if (cn_node_isroot(tn)) {
-            if (!rlen_node || len > rlen) {
-                rlen_node = tn;
-                rlen = len;
-            }
-        } else {
-            if (!ilen_node || len > ilen) {
-                ilen_node = tn;
-                ilen = len;
+        }
+        rmlock_runlock(lock);
+
+        if (len > 0)
+            continue;
+
+        /* We currently have no way to delete the rightmost node if it is
+         * empty because the work scheduling and compaction job apparatus
+         * requires at least one kvset in a node.  Until that changes, we
+         * can look for and remove all rightmost empty nodes periodically
+         * as we perform this shape check.
+         */
+        rmlock_wlock(&tree->ct_lock);
+        tn = list_last_entry_or_null(&tree->ct_nodes, typeof(*tn), tn_link);
+
+        while (tn && cn_ns_kvsets(&tn->tn_ns) == 0 && tree->ct_fanout > 1) {
+            struct cn_tree_node *left = list_prev_entry(tn, tn_link);
+
+            /* We can only remove the rightmost node if neither it nor
+             * its left neighbor are undergoing a spill (required to
+             * correctly coordinate with incremental spill).
+             */
+            mutex_lock(&tree->ct_ss_lock);
+            if (tn->tn_ss_spilling || left->tn_ss_spilling) {
+                tn = NULL;
+            } else {
+                struct route_map *map = tree->ct_route_map;
+                uint8_t kbuf[HSE_KVS_KEY_LEN_MAX];
+                merr_t err;
+
+                memset(kbuf, -1, sizeof(kbuf));
```

Comment:
Would it be optimal to memset this buffer outside the critical sections? That would also enable us to reuse this buffer when there are several empty nodes...

Reply:
True, but I was trying to minimize the amount of clutter in blocks that didn't need it :-)
Lemme take a fresh look...

Reply:
I guessed that should have been the reason :-) But still thought it's worth pointing out as we are holding the tree lock in write mode here...

Reply:
Yeah, I know :-)  I just figured this is really rare..  I moved it up...

Reply:
Done.